src/contents/private-smart-contracts-with-solidity-and-circom.mdx

```diff
@@ -0,0 +1,370 @@
+git clone https://github.com/iden3/circom.git
+cd circom
+cargo build --release
+cargo install --path circom
+npm install -g snarkjs
+```
+
+We'll also use the Circom libraries where the Poseidon function that we'll be using is located.
+
+```bash
+git clone https://github.com/iden3/circomlib.git
+```
+
+## 1. Public key creation
+
+The method we will use to conduct anonymous and secure voting is by proving that we are part of a group without revealing our identity. For example, I will vote for the president of Honduras, demonstrating that I am a Honduran without revealing which specific Honduran I am. This is called "proof of inclusion in a set."
+
+The most practical way to achieve this in zk and blockchain is through Merkle trees. We will place the voters as leaves in the tree and prove that we are one of them without disclosing which one.
+
+Since the tree is public, we will use a set of public-private key pairs so that each voter can cast their vote only once.
+
+You might wonder if we can use the public keys from our Ethereum wallet (e.g., from MetaMask). In future guides like this one, I'll address that topic just as I did with noir. To reach that point, you'll need the fundamentals from this guide. So stay tuned and subscribe!
+
+Now, let's create the public keys for the following private keys using the circuit privateKeyHasher.circom below:
+* `111`
+* `222`
+* `333`
+* `444`
+
+`privateKeyHasher.circom`
+```cpp
+pragma circom 2.0.0;
+
+include "circomlib/circuits/poseidon.circom";
+
+template privateKeyHasher() {
+    signal input privateKey;
+    signal output publicKey;
+    component poseidonComponent;
+    poseidonComponent = Poseidon(1);
+    poseidonComponent.inputs[0] <== privateKey;
+    publicKey <== poseidonComponent.out;
+    log(publicKey);
+}
+
+component main = privateKeyHasher();
+```
+
+`input.json`
+```json
+{
+    "privateKey": "111"
+}
+```
+
+Compile and compute the circuit with the commands below, and you'll see the result in the terminal.
+
+```bash
+circom privateKeyHasher.circom --r1cs --wasm --sym --c
+node privateKeyHasher_js/generate_witness.js privateKeyHasher_js/privateKeyHasher.wasm input.json witness.wtns
+```
+
+The result of the 4 private keys should be as follows:
+
+| Private key | Public key |
+| -------- | ------- |
+| 111  | 13377623690824916797327209540443066247715962236839283896963055328700043345550    |
+| 222 | 3370092681714607727019534888747304108045661953819543369463810453568040251648     |
+| 333    | 19430878135540641438890585969007029177622584902384053006985767702837167003933    |
+| 444    | 2288143249026782941992289125994734798520452235369536663078162770881373549221    |
+
+Is it necessary to do this through Circom? The answer is no. By using Circom, we're performing a lot of unnecessary computation. For now, we're doing it this way to ensure that the implementation of the Poseidon hashing algorithm we'll use later is compatible. This is not recommended for production projects.
```

Comment:
Maybe leave a comment here what your recommendations are for production projects?